File: Iteration1/Algorithm/Sliding Window/longest-substring-without-repeating-characters.cs
using System.Linq;
using System.Collections.Generic;

public class Solution {
    public int LengthOfLongestSubstring(string s) {
        if (s.Length < 2) {
            return s.Length;
        }
        
        
        var a = s.ToCharArray();
        var dic = new Dictionary<char, int>();
        
        int max = 0;
        int currentLength = 0;
        
        for (int i = 0; i < a.Length; i++) {
            var lastIndex = dic.ContainsKey(a[i]) ? dic[a[i]] : -1;
            
            if (lastIndex == -1) {
                currentLength++;
                

            }
            else {
                currentLength = System.Math.Min(i - lastIndex, currentLength + 1);
            }
            dic[a[i]] = i;
            if (currentLength > max) {
                max = currentLength;
            }
        }
        
        return max;
    }
}

File: 1000 and 1 tasks/new-top-150/longest-substring-without-repeating-characters.java
import java.util.Arrays;
import java.util.Collections;
import java.util.HashMap;
import java.util.Map;

class Solution {
    public int lengthOfLongestSubstring(String s) {
        var maxLength = 0;
        var charTable = new HashMap<Character, Integer>();

        var left = 0;
        for (var right = 0; right < s.length(); right++) {
            var c = s.charAt(right);
            if (!charTable.containsKey(c) || charTable.get(c) < left) {
                maxLength = Math.max(maxLength, right - left + 1);
            } else {
                left = charTable.get(c) + 1;
            }
            charTable.put(c, right);
        }

        return maxLength;
    }
}